### ISSUE TITLE
Exploring memif socket proxy for direct memif memif

### ISSUE THREAD

**Issue by edwarnicke** [networkservicemesh/networkservicemesh#570]
## Intro to the idea

For the moment we've hit the wall on direct memif memif via bind mounts due to limitations in MountPropogation policy for Mounts returned by DevicePlugin API.

So we try another idea :)

Memif fundamentally works by passing messages back and forth from the slave to the master over a unix file socket.  Most of these messages are 'messages', but some of them are file descriptors being passed over the unix file socket for shared memory regions.

In principle we could put a memif proxy in the middle, provided in the vppagent-dataplane such that

```
slave <---> nsc-socket <----> proxy <---- nse-socket ----> master
```

Basically, the 'proxy' listens on the nsc-socket (the one mapped into the nsc's workspace) for incoming connections from the slave.  When it receives one, it opens a connection via the nse-socket (the one mapped into the nse's workspace) to the master.  The proxy then passes all messages from the slave to the master, and from the master to the slave, including transferred file descriptors.

**Comment by edwarnicke** [comment]
## Hopefully helpful pointers on memif

[There is a super helpful slide deck on memif in Google Docs](https://docs.google.com/presentation/d/1KrpOUUD7oHML6Plge_4E3-oq5YBun5T8pfoDiCBVk74/edit#slide=id.g2ada28b8cd_0_12).  And there's a helpful [MD file for libmemif](https://github.com/FDio/vpp/blob/master/extras/libmemif/libmemif_doc.md).

It's probably important to know that memif connections are going to be AF_UNIX and SOCK_SEQPACKET.

For most messages, its super simple, pass them back and forth transparently.  For ADD_REGION, things will be trickier, because it will involve the proxy receiving an fd from the master, and passing it through to the slave.  It is probably the case that the proxy should close the fd after passing it on to the slave.

**Comment by edwarnicke** [comment]
Implementation suggestions:  write a library in Go, and start experimenting with it as a standalone app.
It is not yet clear to me whether the proxy will live inside the vpp-agent runtime or as a separately spawned runtime in the container.

[There are examples of using Go to pass fds](https://stackoverflow.com/questions/47644667/is-it-possible-to-use-go-to-send-and-receive-file-descriptors-over-unix-domain-s).

**Comment by nickolaev** [comment]
@lobkovilya is this done already?
